File: contest/joi2011ho/a/main.py
# import bisect
# import collections
# import copy
# import heapq
import itertools

# import math
# import string
import sys

# import numpy


def inp_i():
    return int(sys.stdin.readline().rstrip())


def inp_li():
    return list(map(int, sys.stdin.readline().rstrip().split()))


def inp_s():
    return sys.stdin.readline().rstrip()


def inp_ls():
    return list(sys.stdin.readline().rstrip().split())


def main():
    M, N = inp_li()
    K = inp_i()
    area = [list(inp_s()) for _ in range(M)]

    j_n_acc = [
        [0] + list(itertools.accumulate(map(lambda x: 1 if x == "J" else 0, lst))) for lst in area
    ]
    o_n_acc = [
        [0] + list(itertools.accumulate(map(lambda x: 1 if x == "O" else 0, lst))) for lst in area
    ]
    # j_m_acc = list(zip(*[[0] + list(itertools.accumulate(lst)) for lst in zip(*j_n_acc)]))
    # o_m_acc = list(zip(*[[0] + list(itertools.accumulate(lst)) for lst in zip(*o_n_acc)]))
    j_mn_acc = [[0] * (N + 1) for _ in range(M + 1)]
    o_mn_acc = [[0] * (N + 1) for _ in range(M + 1)]
    for m in range(1, M + 1):
        for n in range(1, N + 1):
            j_mn_acc[m][n] += j_mn_acc[m - 1][n] + j_n_acc[m - 1][n]
            o_mn_acc[m][n] += o_mn_acc[m - 1][n] + o_n_acc[m - 1][n]

    for _ in range(K):
        ans1 = ans2 = 0
        a, b, c, d = inp_li()
        ans1 = j_mn_acc[c][d] - j_mn_acc[a - 1][d] - j_mn_acc[c][b - 1] + j_mn_acc[a - 1][b - 1]
        ans2 = o_mn_acc[c][d] - o_mn_acc[a - 1][d] - o_mn_acc[c][b - 1] + o_mn_acc[a - 1][b - 1]
        print(ans1, ans2, (c - a + 1) * (d - b + 1) - ans1 - ans2)


if __name__ == "__main__":
    main()

File: contest/joi2011ho/a/main.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric> // accumulate関数のために必要
using namespace std;

int inp_i() {
    int x;
    cin >> x;
    return x;
}

vector<int> inp_li() {
    int n;
    cin >> n;
    vector<int> lst(n);
    for (int i = 0; i < n; ++i)
        cin >> lst[i];
    return lst;
}

string inp_s() {
    string s;
    cin >> s;
    return s;
}

vector<string> inp_ls() {
    int n;
    cin >> n;
    vector<string> lst(n);
    for (int i = 0; i < n; ++i)
        cin >> lst[i];
    return lst;
}

int main() {
    int M = inp_i(), N = inp_i();
    int K = inp_i();
    vector<string> Area(M);
    for (int i = 0; i < M; ++i)
        Area[i] = inp_s();
    
    vector<vector<int>> J(M, vector<int>(N + 1, 0));
    vector<vector<int>> O(M, vector<int>(N + 1, 0));
    for (int i = 0; i < M; ++i) {
        for (int j = 0; j < N; ++j) {
            if (Area[i][j] == 'J')
                J[i][j + 1] = J[i][j] + 1;
            else
                J[i][j + 1] = J[i][j];
            
            if (Area[i][j] == 'O')
                O[i][j + 1] = O[i][j] + 1;
            else
                O[i][j + 1] = O[i][j];
        }
    }

    for (int k = 0; k < K; ++k) {
        int ans1 = 0, ans2 = 0;
        int a, b, c, d;
        cin >> a >> b >> c >> d;
        for (int m = a - 1; m < c; ++m) {
            ans1 += J[m][d] - J[m][b - 1];
            ans2 += O[m][d] - O[m][b - 1];
        }
        cout << ans1 << " " << ans2 << " " << (c - a + 1) * (d - b + 1) - ans1 - ans2 << endl;
    }

    return 0;
}
